src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -309,57 +271,114 @@ instance
 -- We can only impose some restrictions on @n@ and @m@.
 --
 instance
-  ( KnownNat m
+  ( Symbolic (Interpreter (Zp p))
+  , KnownNat m
   , KnownNat k
   , Mod k m ~ 0
-  , Finite (Zp p)
   ) => Concat (ByteString m (Interpreter (Zp p))) (ByteString k (Interpreter (Zp p))) where
 
     concat = fromConstant @Natural . foldl (\p y -> toConstant y + p `shift` m) 0
         where
             m = Haskell.fromIntegral $ value @m
 
+instance
+  ( Mod k m ~ 0
+  , (Div k m) * m ~ k
+  , Arithmetic a
+  ) => Concat (ByteString m (ArithmeticCircuit a)) (ByteString k (ArithmeticCircuit a)) where
+
+    concat bs = ByteString $ bsCircuit {acOutput = bsOutputs}
```

Comment:
Try to use functions from `Package` class here

Reply:
(I think many of the "pure" implementations defined here are actually harder to understand than their circuit counterparts, I think we can just generalize the circuit ones to `Symbolic` interface and remove the pure implementations)